File: java/src/ch22/UnionFind.java
package ch22;

import java.util.HashMap;
import java.util.Map;

public class UnionFind {
    private Map<Integer, Integer> fathers;
    private Map<Integer, Integer> sizeOfSets;
    private int numberOfSets;

    public UnionFind() {
        fathers = new HashMap<>)();
        sizeOfSets = new HashMap<>();
    }

    public void merge(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            fathers.put(rootX, rootY);
            numberOfSets--;
            sizeOfSets.put(rootY, sizeOfSets.get(rootX) + sizeOfSets.get(rootY));
        }
    }

    private int find(int x) {
        int node = x;
        while (fathers.containsKey(node)) {
            node = fathers.get(node);
        }
        int root = node;
        
        while (x != root) {
            int currentFather = fathers.get(x);
            fathers.put(x, root);
            x = currentFather;
        }
        return root;
    }

    private void add(int x) {
        if (!fathers.containsKey(x)) {
            fathers.put(x, null);
            sizeOfSets.put(x, 1);
            numberOfSets++;
        }
    }

    public boolean isConnected(int x, int y) {
        return find(x) == find(y);
    }

    public int getNumberOfSets() {
        return numberOfSets;
    }

    public int getSizeOfSets(int x) {
        return sizeOfSets.get(find(x));
    }
}

File: python/ch22-unionfind/union_find.py
class UnionFind:
    
    def __init__(self):
        self.fathers = {}
        self.size_of_sets = {}
        self.number_of_sets = 0
        
    def merge(self, x, y):
        root_x, root_y = self.find(x), self.find(y)
        if root_x != root_y:
            self.fathers[root_x] = root_y
            self.number_of_sets -= 1
            self.size_of_sets[root_y] += self.size_of_sets[root_x]          
    
    def find(self, x):
        root = x
        while self.fathers[root]:
            root = self.fathers[root]
            
        while x != root:
            original_father = self.fathers[x]
            self.fathers[x] = root
            x = original_father
            
        return root
    
    def add(self, x):
        # if self.fathers[x] == None:
        if not x in self.fathers:
            self.fathers[x] = None
            self.size_of_sets[x] = 1
            self.number_of_sets +=1
            
    def is_connected(self, x, y):
        return self.find(x) == self.find(y)
    
    def get_number_of_sets(self):
        return self.number_of_sets()
    
    def get_size_of_sets(self, x):
        return self.size_of_sets[self.find(x)]
    